### ISSUE TITLE
[BUG] Custom Volt functions not working inside macros

### ISSUE THREAD

**Issue by integry** [phalcon/cphalcon#1176]
It happens when the custom function is called through $this (the Volt instance).

For example, I set up this function:
```php
$this->_compiler->addFunction('t', function($resolvedArgs, $exprArgs) {
	return '$this->getDI()->get(\'application\')->translate(' . $resolvedArgs . ')';
});
```

It works fine as long as it's not used inside a macro, for example:
```
{%- macro textfld(field, title) %}
        <label>{{ t(title) }}</label>
        {{ text_field(field }}
{%- endmacro %}
```

Calling the t() function will produce this error: Fatal error: Using $this when not in object context

It happens because the macros are compiled into PHP functions and calling them loses the object context. This could be resolved by passing the Volt instance (for example, as *$volt*) as a parameter to the compiled macro function and making the Volt instance available via $volt in the general template scope as well. Then in the custom functions we could use $volt instead of $this.

My current workaround is to access the DI via $GLOBALS['di'] in the compiler function, but this feels so wrong.

**Comment by nurtext** [comment]
Can and will this be fixed in 2.0?

**Comment by andresgutierrez** [comment]
This is fixed in Phalcon 2.0.4
